Enable Arrays to be marked as required, and fixes for strings marked as required

crumblix opened the pull request:
### Reasons for making this change
Reasons for making this change

New feature:

Enable the arrays to be marked as "required". Meaning that the user must select "at least one" option from the list.

This references the array part of ticket:
#411

It also refers to the string part of
#410 and #402

### Checklist

* [x] **I'm updating documentation**
  - [x] I've checked the rendering of the Markdown text I've added
  - [ ] If I'm adding a new section, I've updated the Table of Content
* [x] **I'm adding or updating code**
  - [ ] I've added and/or updated tests
  - [x] I've updated docs if needed
* [x] **I'm adding a new feature**
  - [x] I've updated the playground with an example use of the feature

MoOx commented on README.md:
````diff
@@ -558,6 +558,8 @@ const uiSchema = {
 };
 ```
 
+In addition it is valid to maked an array as "required".  This means the user must select at least one items to add to the array.
````
to maked? Looks like a typo. Maybe "to mark"?

crumblix commented:
Nice pickup ... thank you!!! :)

n1k0 commented on src/components/fields/ArrayField.js:
```diff
@@ -216,8 +216,12 @@ class ArrayField extends Component {
       this.asyncSetState({
-        items: this.state.items.filter((_, i) => i !== index)
+        items: newitems
       }, {validate: true}); // refs #195
```
```js
const items = this.state.items.filter((_, i) => i !== index);
this.asyncSetState({
  items: items.length > 0 ? items : undefined,
}, {validate: true}); // refs #195
```

n1k0 commented on src/components/widgets/BaseInput.js:
```diff
@@ -8,21 +8,25 @@ function BaseInput(props) {
+        return _onChange(undefined);
+      } else {
+        return _onChange(event.target.value);
+      }}}/>
```
```jsx
const _onChange = ({target: {value}}) => {
  return props.onChange(value === "" ? undefined : value);
};
// ...
onChange={_onChange} />
```

n1k0 commented on src/components/widgets/CheckboxesWidget.js:
```diff
@@ -10,7 +10,12 @@ function selectValue(value, selected, all) {
+    return undefined;
+  } else {
+    return ret;    
+  }  
```
```js
const ret = selected.filter(v => v !== value);
return ret.length === 0 ? undefined : ret;
```

n1k0 commented on src/components/widgets/CheckboxesWidget.js:
```diff
@@ -19,7 +24,7 @@ function CheckboxesWidget(props) {
     <div className="checkboxes" id={id}>{
       enumOptions.map((option, index) => {
-        const checked = value.indexOf(option.value) !== -1;
+        const checked = (value === undefined) ? false : value.indexOf(option.value) !== -1;
```
Nit: Unnecessary wrapping parenthesis.

n1k0 commented on src/components/widgets/TextareaWidget.js:
```diff
@@ -22,7 +23,11 @@ function TextareaWidget({
+        return _onChange(undefined);
+      } else {
+        return _onChange(event.target.value);
+      }}}/>
```
See previous impl. suggestion above.

crumblix commented on README.md:
````diff
@@ -558,6 +558,8 @@ const uiSchema = {
 };
 ```
 
+In addition it is valid to mark an array as "required".  This means the user must select at least one item to add to the array.
````
Fair call.  I wasn't aware of that directive.  If we can achieve a similar effect with existing code that is standards compliant I would certainly support that.

MoOx replied:
Indeed that's a serious diff. Not sure I am comfortable with that but I guess we can't have the silver bullet.

crumblix replied:
The SelectWidget is an interesting case.  In truth I don't use it much because I've written my own custom widget based on react-select which better suited my needs for a nullable and multiselect-able drop down widget.  (Happy to share if anyone wants it, but I thought since it was an extra dependency it was better off outside the core project.)

Although it flies in the face of this change, I would tend to opt the other direction for the SelectWidget ... if only because the programmer has explicitly set the valid return values so should be expecting it, so perhaps we don't need to try and interpret their intent.

olzraiti replied:
This wouldn't affect my code base, but I'm not sure if we would benefit enough from adding this considering the consequences.

We aren't extending the json schema spec in a way that would make something possible that was impossible before (using `minItem` provides the functionality that is wanted here). We are merely making something more convenient than before, with the cost of breaking compatibility with json schema spec.

crumblix replied:
Yep.  More than happy to leave the array case out.  If we can achieve the same result using minItem and it's more standards compliant I can't see why we wouldn't.

maartenth replied:
I am in favor of this change. It is consequent with the empty string to undefined functionality. The error message you get when you don't select a value is then always the same, whether you never selected a value or you selected and then deselected a value. If something looks the same in the ui, I think it should be the same.

For JSON Schema, the empty array is of course a valid array, but JSON Schema is not a form builder and the difference between an empty array and undefined is simply not visible in a form and should as a consequence not be settable through the form, in my opinion.

n1k0 replied:
> JSON Schema is not a form builder

The issue is that people import existing schemas and expect them to work with forms with no need to tweak their schemas (uiSchemas have been introduced for that specific purpose). 

That could be addressed with an appropriate FAQ/Caveats section though. The same way we could document the `minItems` jsonschema directive.

> the difference between an empty array and undefined is simply not visible in a form

That's indeed a fair point. Another approach would be to make the difference actually visible, but that's rather tricky to get right, and would probably clutter UX a lot. Also this would have to work with bootstrap and other styles.

So my current thinking is:

1. Keep the *empty string as undefined value* feature from this patch.
1. Remove the *required array should contain at least one item* one and add proper docs describing how to achieve it using the `minItems` jsonschema directive.
1. Leave `SelectWidget` as it is, users should provide enum values matching their expectations.

What do people think?

dehli replied:
I agree with you @n1k0 . I'm currently using minItems in my application b/c that's the `json-schema` way of doing it. If we added documentation describing it, then it'll be more obvious that it's the way to do it.

olzraiti replied:
Sounds good to me! Empty string as undefined seems well-grounded.

maartenth replied:
I'm OK with your proposed solution, @n1k0. I think it is an improvement to the current situation either way.

n1k0 replied:
Good, @crumblix would you mind updating the patch accordingly if you agree with this decision? :)

Thanks a bunch :+1:

crumblix commented:
Thank you very much for the code review!  Much appreciated, I have no problem addressing a few nits :)

dehli commented on src/components/fields/ArrayField.js:
```diff
@@ -216,8 +216,9 @@ class ArrayField extends Component {
       if (event) {
         event.preventDefault();
       }
+      var items = this.state.items.filter((_, i) => i !== index);
```
I think should be a `const` or `let`. It looks like you never assign to it again, so a `const` would be the way to go.

crumblix commented:
Thanks again everyone for the feedback :)
I reverted the documentation and fixed the last nit to make the code cleaner and better.

I'll need to read up on minItems before making any changes to the documentation for that as I haven't used it before.

My plan is to try and get it working in the playground later before updating the documentation about it.

Either that or if someone who knows how it works already sends me a sentence I'll just include it.

crumblix commented:
OK, so I've reverted the array changes, played around with minItems in the playground, and tried to make it clear how to achieve a similar (and more feature rich) result by using it in the documentation.

Thanks for pointing out minItems!

n1k0 commented on README.md:
````diff
@@ -547,6 +547,23 @@ const uiSchema = {
 };
 ```
 
+Unlike other types, array's do not support the `required` property.  You can however instead specify the minimum number of items the user must select with the `minItems` property.
````
It actually supports the `required` property, it only checks that an array property value is provided, which may or may not be empty. Put another way, for

```js
schema = {
  type: "object",
  required: ["foo"],
  properties: {
    foo: {type: "array"}
  }
}
```

`{foo: []}` is valid against this schema while `{}` is not.

crumblix replied:
No problem!  While it's true supports the required property, I suspect that it doesn't operate in the fashion most UI builders would expect, at the very least it's a potential source of confusion.  I'd try and update the documentation to be clear in that regard. :)

And yes to the tests as well! :)

n1k0 replied:
> I suspect that it doesn't operate in the fashion most UI builders would expect

I think we need to stay true to the jsonschema spec as much as possible, as it's our data model layer behind our form views. One of the goal of this lib is also to make forms and their underlying data model more robust and consistent. I wouldn't want to bring regular web forms flakiness to jsonschema, which would kind of defeat its purpose if you ask me :)

n1k0 commented on README.md:
````diff
@@ -547,7 +547,7 @@ const uiSchema = {
 ```
 
-Unlike other types, array's do not support the `required` property.  You can however instead specify the minimum number of items the user must select with the `minItems` property.
+Care should be taken when using the `required` property with arrays.  An empty array is sufficient to pass that validation check.  If you wish to ensure the user populates the array, you can specify the minimum number of items the user must select with the `minItems` property.
````
This is perfect :+1:

crumblix replied:
Excellent :)
I'll have to get to the tests over the weekend.  It's a part of the code base I haven't been into yet.

crumblix replied:
Unfortunately there wasn't a text area widget section that I could see so I created one .... but it isn't right.  I will have to look at it over the weekend.  

What's checked in is passing, but it's not using the widget.

I've tried the change:

    const uiSchema = {"ui:widget": "textarea"};

    it("should handle an empty string change event", () => {
      const {comp, node} = createFormComponent({schema: {
        type: "string",
      }, uiSchema});

... which seems to me like it should work but that fails for reasons that currently escape me.  I'll need to look at it over the weekend.  Got to head off now though!

crumblix commented:
You're welcome :), but I think Textarea widget test is wrong ... isn't it?

n1k0 commented:
> I think Textarea widget test is wrong ... isn't it?

It was, fixed by 8c4707a125bc191c9ac3bf6c39d706062a26aa2b :)

n1k0 commented:
Released in [v0.42.0](https://github.com/mozilla-services/react-jsonschema-form/releases/tag/v0.42.0).
